
Привет, меня зовут Головатский Николай, я буду проверять твой проект.

В Практикуме принято общаться на "ты", однако если это неудобно - дай знать, и можем перейти на более официальное общение.

Ниже ты найдешь мои комментарии - **пожалуйста, не перемещай, не изменяй и не удаляй их,** это может усложнить и затянуть повторную проверку. 

Я буду использовать цветовую разметку: 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Если все сделано хорошо.</div>

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Некритичное замечание, реккомендация на будущее. Для принятия проекта исправление не обязательно.</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Критичная ошибка, необходимо исправить, иначе проект не будет принят.
</div>

Давай работать над проектом в диалоге: если ты что-то меняешь в проекте или отвечаешь на мои комменатри — пиши об этом. Мне будет легче отследить изменения, если ты выделишь свои комментарии:
Например так:
<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Hello World!
</div>

Если писать особо нечего, просто оставляй комментарий с галочкой 

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
☑️
</div>

<div class="alert alert-block alert-success">
<b>Общее впечатление:</b> 
    
- Спасибо за работу :)
- Определенно могу сказать, что проект очень сильный;
- Мне очень понравилось, как ты детально погружаешься в данные и не упускаешь никаких мелочей;
- В проекте осталось пара моментов, после правки которых, я смогу принять работу😉
    
</div>

Изменение

Изменение

# Исследование надежности заемщиков.

**Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов. Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.**

**Цель исследования** — ответить на вопросы:

**1. Есть ли зависимость между количеством детей и возвратом кредита в срок?**
**2. Есть ли зависимость между семейным положением и возвратом кредита в срок?**
**3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**
**4. Как разные цели кредита влияют на его возврат в срок?**

**Ход исследования**

**Данные о платёжеспособности клиентов мы получили из файла /datasets/data.csv. О качестве данных ничего не известно. Поэтому перед ответом на вопросы понадобится обзор данных.**

**Проверим данные на ошибки и оценим их влияние на исследование. На этапе предобработки данных исправим самые критичные ошибки данных.**

**Таким образом, исследование пройдёт в три этапа:**
**1. Обзор данных.**
**2. Предобработка данных.**
**3. Ответы на вопросы.**

### Шаг 1. Обзор данных

In [1]:
import pandas as pd # импорт библиотеки pandas

Проверим данные на ошибки конструкцией `try-except`:

In [2]:
try:
    df = pd.read_csv('/datasets/data.csv') # чтение файла с данными и сохранение в df
except:
    df = pd.read_csv('/datasets/data.csv')

In [3]:
df.head(10) # получение первых 10 строк таблицы df

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

В столбце `days_employed` есть отрицательные данные.

In [4]:
df.info()# получение общей информации о данных в таблице df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


В таблице двенадцать столбцов. Тип данных в столбцах — `object`, `int64`, `float64`, что соответсвует данным каждого столбца.
Для удобства исследования, изменим тип данных в столбце `total_income` с `float64` на `int64`.

Согласно документации к данным:

* `children` — количество детей в семье
* `days_employed` — общий трудовой стаж в днях
* `dob_years` — возраст клиента в годах
* `education` — уровень образования клиента
* `education_id` — идентификатор уровня образования
* `family_status` — семейное положение
* `family_status_id` — идентификатор семейного положения
* `gender` — пол клиента
* `income_type` — тип занятости
* `debt` — имел ли задолженность по возврату кредитов
* `total_income` — ежемесячный доход
* `purpose` — цель получения кредита

Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.

**Выводы**

В каждой строке таблицы — данные о клиентах. Часть колонок описывает личную информацию о клиентах: дети, стаж, образование, семейное положение, пол, тип занятости. Другая часть описывает финансовую составляющую: задолженность по кредитам, доход, цель получения кредита. 

Данных достаточно для ответа на поставленные вопросы. Но встречаются пропуски в данных.

Изменим тип данных. Обработаем столбец с аномалиями. Устраним пропуски в даных. 

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Хорошее начало, информация о данных изучена👍</div>

## Предобработка данных

* Обработаем столбец с аномалиями.
* Исключим пропуски. 
* Изменим тип даных.
* Проверим данные на дубликаты.

### Шаг 2.1 Заполнение пропусков

Посчитаем, сколько в таблице пропущенных значений. Для этого применим методы `pandas`: `isna()`, `sum()`

In [5]:
df.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

Мы обнаружили пропуски в столбцах: `days_employed` и `total_income`. 
Пропусков в двух столбцах одинаковое количество. Посчитаем долю пропусков в общем количестве строк: (2174 / 21525)* 100%, получим 10,1 %. Достаточно высокий процент. Для заполнения пропущенных строк попытаемся понять есть ли взаимосвязи между пропусками.

In [6]:
df[df['days_employed'].isna()].head(10) # применим метод isna()

children  days_employed  dob_years education  education_id  \
12         0            NaN         65   среднее             1   
26         0            NaN         41   среднее             1   
29         0            NaN         63   среднее             1   
41         0            NaN         50   среднее             1   
55         0            NaN         54   среднее             1   
65         0            NaN         21   среднее             1   
67         0            NaN         52    высшее             0   
72         1            NaN         32    высшее             0   
82         2            NaN         50    высшее             0   
83         0            NaN         52   среднее             1   

            family_status  family_status_id gender  income_type  debt  \
12       гражданский брак                 1      M    пенсионер     0   
26        женат / замужем                 0      M  госслужащий     0   
29  Не женат / не замужем                 4      F    пенсионер     0   
41        женат / замужем                 0      F  госслужащий     0   
55       гражданский брак                 1      F    пенсионер     1   
65  Не женат / не замужем                 4      M    компаньон     0   
67        женат / замужем                 0      F    пенсионер     0   
72        женат / замужем                 0      M  госслужащий     0   
82        женат / замужем                 0      F    сотрудник     0   
83        женат / замужем                 0      M    сотрудник     0   

    total_income                                purpose  
12           NaN                        сыграть свадьбу  
26           NaN                            образование  
29           NaN       строительство жилой недвижимости  
41           NaN       сделка с подержанным автомобилем  
55           NaN                        сыграть свадьбу  
65           NaN  операции с коммерческой недвижимостью  
67           NaN                покупка жилья для семьи  
72           NaN  операции с коммерческой недвижимостью  
82           NaN                                  жилье  
83           NaN                                  жилье

**Вывод**
Столбцы: `days_employed` и `total_income` имеют одиноковые пропуски, но они не связаны с другими данными (возраст, пол, семейное положение, профессия и тд.).
Взаимосвязи между пропусками и какой-либо другой информацией нет.
Можно сделать вывод, что пропуски данных о трудовом стаже и доходе взаимосвязаны только между собой. Так как эти данные важны для исследования, будем искать способы заполнить данные пропуски.

Данные пропуски относятся к количественным переменным, к ним применим методы среднее арифметическое `mean()` или медиана `median()`.
В случае ежемесячного дохода и общего стажа лучше применить метод `median()`. Он покажет более точные данные, чем метод `mean()`, так как среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства. Для столбца `total_income` применим метод `groupby` в конбинации с функцией `transform`, которая поможет заполнить медианными значениями в разрезе `income_type`. 

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Привет, Николай! Нашла метод transform и решила применить его в проекте. Спасибо за идею и мои новые знания)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> 👍 </div>

In [7]:
median_table = df.groupby('income_type')['total_income'].transform('median') # применим метод median() с функцией transform() 
df['total_income'] = df['total_income'].fillna(median_table) # выведим данные для просмотра
display(df.head(10))

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0    -926.185831         27    высшее             0   
6         0   -2879.202052         43    высшее             0   
7         0    -152.779569         50   СРЕДНЕЕ             1   
8         2   -6929.865299         35    ВЫСШЕЕ             0   
9         0   -2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

В столбце `days_employed` обнаружили отрицательные значения. В данном столбце хранится информация об общем трудовом стаже в днях. Данная информация не может быть отрицательной. Скорее всего при выгрузке данных произошла техническая ошибка. Исправим отрицательные данные на положительные. Применим метод `abs`.

In [8]:
df['days_employed'] = df['days_employed'].abs()
display(df.head(10)) # выведим данные для просмотра

children  days_employed  dob_years education  education_id  \
0         1    8437.673028         42    высшее             0   
1         1    4024.803754         36   среднее             1   
2         0    5623.422610         33   Среднее             1   
3         3    4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   
5         0     926.185831         27    высшее             0   
6         0    2879.202052         43    высшее             0   
7         0     152.779569         50   СРЕДНЕЕ             1   
8         2    6929.865299         35    ВЫСШЕЕ             0   
9         0    2188.756445         41   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   
5  гражданский брак                 1      M   компаньон     0  255763.565419   
6   женат / замужем                 0      F   компаньон     0  240525.971920   
7   женат / замужем                 0      M   сотрудник     0  135823.934197   
8  гражданский брак                 1      F   сотрудник     0   95856.832424   
9   женат / замужем                 0      M   сотрудник     0  144425.938277   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

In [9]:
median_days_employed = df['days_employed'].median() # применим метод median() 
display(median_days_employed) # выведим данные для просмотра

2194.220566878695

Заменим пропуски в столбце `days_employed`.

In [10]:
df['days_employed'] = df['days_employed'].fillna(value=median_days_employed) # применим метод fillna()
display(df['days_employed'].head(10)) # выведим данные для просмотра

0      8437.673028
1      4024.803754
2      5623.422610
3      4124.747207
4    340266.072047
5       926.185831
6      2879.202052
7       152.779569
8      6929.865299
9      2188.756445
Name: days_employed, dtype: float64

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> 
Есть еще один, более интересный способ заполнения пропусков. Например, доход клиента (total_income) можно заполнить медианными значениями в разрезе income_type - для каждого income_type посчитать медианный total_income и заполнить пропуски ими. Так мы не усредним данные и меньше повлияем на распределение)

### Шаг 2.2 Проверка данных на аномалии и исправления.

Проверим столбцы: `children`, `dob_years`, `education_id`, `family_status_id`, `gender`, `debt` на наличие аномалий.

Проверим столбец `children`. Применим метод `value_counts()`.

In [11]:
df['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

В данном столбце видим следующие аномалии: отрицательное значение и количество детей 20 у 76 клиентов. Скорее всего это ошибка вызвана человеческим фактором, т.е. при заполнении данных, произошла ошибка. Допустим, что детей не 20, а 2.
Произведем следующие действия:
1. Заменим -1 на 1;
2. Исправим 20 на 2;
3. Проверим результат.

Для этого воспользуемся методом `loc`.

In [12]:
df.loc[df['children']==-1, 'children']=1
df.loc[df['children']==20, 'children']=2
df['children'].value_counts()

0    14149
1     4865
2     2131
3      330
4       41
5        9
Name: children, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Согласен😉</div>

Теперь проверим столбец `dob_years`. Применим метод `value_counts()`.

In [13]:
df['dob_years'].value_counts()

35    617
40    609
41    607
34    603
38    598
42    597
33    581
39    573
31    560
36    555
44    547
29    545
30    540
48    538
37    537
50    514
43    513
32    510
49    508
28    503
45    497
27    493
56    487
52    484
47    480
54    479
46    475
58    461
57    460
53    459
51    448
59    444
55    443
26    408
60    377
25    357
61    355
62    352
63    269
64    265
24    264
23    254
65    194
66    183
22    183
67    167
21    111
0     101
68     99
69     85
70     65
71     58
20     51
72     33
19     14
73      8
74      6
75      1
Name: dob_years, dtype: int64

Выявили 0 в 101 случаи. Сложно понять какая была ошибка, поэтому воспользуемся методом `median()`. Он покажет более точные данные, чем метод `mean()`, так как среднее значение некорректно характеризует данные, когда некоторые значения сильно выделяются среди большинства.

In [14]:
median_dob_years = df['dob_years'].median() # находим медиану 

In [15]:
df.loc[df['dob_years']==0, 'dob_years'] = median_dob_years # методом Loc меняем 0 на полученную медиану
df['dob_years'].value_counts() # проверяем данные

42.0    698
35.0    617
40.0    609
41.0    607
34.0    603
38.0    598
33.0    581
39.0    573
31.0    560
36.0    555
44.0    547
29.0    545
30.0    540
48.0    538
37.0    537
50.0    514
43.0    513
32.0    510
49.0    508
28.0    503
45.0    497
27.0    493
56.0    487
52.0    484
47.0    480
54.0    479
46.0    475
58.0    461
57.0    460
53.0    459
51.0    448
59.0    444
55.0    443
26.0    408
60.0    377
25.0    357
61.0    355
62.0    352
63.0    269
64.0    265
24.0    264
23.0    254
65.0    194
66.0    183
22.0    183
67.0    167
21.0    111
68.0     99
69.0     85
70.0     65
71.0     58
20.0     51
72.0     33
19.0     14
73.0      8
74.0      6
75.0      1
Name: dob_years, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Логично😉</div>

Теперь проверим столбец `education_id`. Применим метод `value_counts()`.

In [16]:
df['education_id'].value_counts()

1    15233
0     5260
2      744
3      282
4        6
Name: education_id, dtype: int64

Теперь проверим столбец `family_status_id`. Применим метод `value_counts()`.

In [17]:
df['family_status_id'].value_counts()

0    12380
1     4177
4     2813
3     1195
2      960
Name: family_status_id, dtype: int64

Теперь проверим столбец `gender`. Применим метод `value_counts()`.

In [18]:
df['gender'].value_counts()

F      14236
M       7288
XNA        1
Name: gender, dtype: int64

Обнаружили XNA. Так как количество таких аномалий составлеяет единицу, на исследование это не повлияет. Удалим данную аномалию.

In [19]:
df = df.loc[df['gender'] != 'XNA'] # применим метод Loc 
df['gender'].value_counts() # проверяем данные

F    14236
M     7288
Name: gender, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Аргументировано👍</div>

Теперь проверим столбец `debt`. Применим метод `value_counts()`.

In [20]:
df['debt'].value_counts()

0    19783
1     1741
Name: debt, dtype: int64

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Молодец, все аномалии изучены и исправлены верно. Двигаемся дальше🔥</div>

### Шаг 2.3. Изменение типов данных.

Воспользуемся методом `astype()`

In [21]:
df['total_income'] = df['total_income'].astype('int') # применим метод astype()
display(df['total_income'].head(10)) # выведим данные для просмотра

0    253875
1    112080
2    145885
3    267628
4    158616
5    255763
6    240525
7    135823
8     95856
9    144425
Name: total_income, dtype: int64

Убедимся, что в таблице не осталось пропусков. Для этого ещё раз посчитаем пропущенные значения.

In [22]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 21524 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21524 non-null  int64  
 1   days_employed     21524 non-null  float64
 2   dob_years         21524 non-null  float64
 3   education         21524 non-null  object 
 4   education_id      21524 non-null  int64  
 5   family_status     21524 non-null  object 
 6   family_status_id  21524 non-null  int64  
 7   gender            21524 non-null  object 
 8   income_type       21524 non-null  object 
 9   debt              21524 non-null  int64  
 10  total_income      21524 non-null  int64  
 11  purpose           21524 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.1+ MB


<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Типы данных исправлены верно, но для их проверки лучше использовать методы info() или dtypes😉</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Поправила 😉
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> 👍 </div>

### Шаг 2.4. Удаление дубликатов.

Для проверки данных на дубликаты необходимо решить какие столбцы могут имень искаженные данные в виде дубликатов. 

Столбцы:
`children` — количество детей в семье, может быть одинаковым у разных клиентов;
`days_employed` — общий трудовой стаж в днях, может быть одинаковый у разных клинтов. У нас были пропуски и мы заменили их медиаными значениями;
`dob_years` — возраст клиента в годах, может быть одинаковым у разных клиентов;
`education` — уровень образования клиента, может быть одинаковым у разных клиентов;
`education_id` — идентификатор уровня образования, может быть одинаковым у разных клиентов;
`family_status` — семейное положение, может быть одинаковым у разных клиентов;
`family_status_id` — идентификатор семейного положения, может быть одинаковым у разных клиентов;
`gender` — пол клиента, может быть одинаковым у разных клиентов;
`income_type` — тип занятости, может быть одинаковым у разных клиентов. У нас были пропуски и мы заменили их медиаными значениями;
`debt` — имел ли задолженность по возврату кредитов, может быть одинаковым у разных клиентов;
`total_income` — ежемесячный доход, может быть одинаковым у разных клиентов;
`purpose` — цель получения кредита, может быть одинаковая у разных клиентов.

Можно сделать вывод, что проверять на явные дубликаты не имеет смысла.

Применим метод поиска неявных дубликатов к следующим столбам: `education`, `family_status`, `income_type`, `purpose`.


Начнем поиск неявных дубликатов со столбца `education`. Применим метод `value_counts()`.

In [23]:
 df['education'].value_counts()

среднее                13750
высшее                  4718
СРЕДНЕЕ                  772
Среднее                  711
неоконченное высшее      667
ВЫСШЕЕ                   274
Высшее                   268
начальное                250
Неоконченное высшее       47
НЕОКОНЧЕННОЕ ВЫСШЕЕ       29
НАЧАЛЬНОЕ                 17
Начальное                 15
ученая степень             4
Ученая степень             1
УЧЕНАЯ СТЕПЕНЬ             1
Name: education, dtype: int64

Видим, что названия отличаются регистром нескольких букв. Применим метод `str.lower()`. И сразу проверим.

In [24]:
df['education'] = df['education'].str.lower()

In [25]:
df['education'].value_counts()

среднее                15233
высшее                  5260
неоконченное высшее      743
начальное                282
ученая степень             6
Name: education, dtype: int64

Теперь проверим столбец `family_status`. Применим метод `value_counts()`.

In [26]:
 df['family_status'].value_counts()

женат / замужем          12380
гражданский брак          4176
Не женат / не замужем     2813
в разводе                 1195
вдовец / вдова             960
Name: family_status, dtype: int64

Теперь проверим столбец `income_type`. Применим метод `value_counts()`.

In [27]:
df['income_type'].value_counts()

сотрудник          11119
компаньон           5084
пенсионер           3856
госслужащий         1459
предприниматель        2
безработный            2
в декрете              1
студент                1
Name: income_type, dtype: int64

Теперь проверим столбец `purpose`. Применим метод `value_counts()`.

In [28]:
df['purpose'].value_counts()

свадьба                                   797
на проведение свадьбы                     777
сыграть свадьбу                           774
операции с недвижимостью                  676
покупка коммерческой недвижимости         664
операции с жильем                         653
покупка жилья для сдачи                   653
операции с коммерческой недвижимостью     651
покупка жилья                             647
жилье                                     647
покупка жилья для семьи                   641
строительство собственной недвижимости    635
недвижимость                              634
операции со своей недвижимостью           630
строительство жилой недвижимости          626
покупка недвижимости                      623
строительство недвижимости                620
покупка своего жилья                      620
ремонт жилью                              612
покупка жилой недвижимости                607
на покупку своего автомобиля              505
заняться высшим образованием      

**Выводы**

Предобработка данных обнаружила четыре проблемы:

- аномалии в столбцах,
- пропущенные значения,
- изменения типа данных,
- неявные дубликаты.

Мы исправили аномалии в тех столбцах, в которых их обнаружили, избавились от неявных дубликатов, изменили тип данных, пропущенные значения заполнили медианными.

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Неявные дубликаты исправлены, но нам необходимо обработать грубые дубликаты строк. Как видно из следующей ячейки, в таблице они присутствуют</div>

In [29]:
# Код ревьюера
df.duplicated().sum()

72

Выявили дубликаты. Методом `drop_duplicates()` удалим дубликаты с изменением индексов и удалением старых.

In [30]:
df = df.drop_duplicates().reset_index(drop=True)
display(df.head(10))

children  days_employed  dob_years education  education_id  \
0         1    8437.673028       42.0    высшее             0   
1         1    4024.803754       36.0   среднее             1   
2         0    5623.422610       33.0   среднее             1   
3         3    4124.747207       32.0   среднее             1   
4         0  340266.072047       53.0   среднее             1   
5         0     926.185831       27.0    высшее             0   
6         0    2879.202052       43.0    высшее             0   
7         0     152.779569       50.0   среднее             1   
8         2    6929.865299       35.0    высшее             0   
9         0    2188.756445       41.0   среднее             1   

      family_status  family_status_id gender income_type  debt  total_income  \
0   женат / замужем                 0      F   сотрудник     0        253875   
1   женат / замужем                 0      F   сотрудник     0        112080   
2   женат / замужем                 0      M   сотрудник     0        145885   
3   женат / замужем                 0      M   сотрудник     0        267628   
4  гражданский брак                 1      F   пенсионер     0        158616   
5  гражданский брак                 1      M   компаньон     0        255763   
6   женат / замужем                 0      F   компаньон     0        240525   
7   женат / замужем                 0      M   сотрудник     0        135823   
8  гражданский брак                 1      F   сотрудник     0         95856   
9   женат / замужем                 0      M   сотрудник     0        144425   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу  
5               покупка жилья  
6           операции с жильем  
7                 образование  
8       на проведение свадьбы  
9     покупка жилья для семьи

Проверим получившийся результат

In [31]:
df.duplicated().sum()

0

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Николай, я не могу понять какие дубликаты нужно удалить. При проверке на дубликаты показывает ,что все 72 дубликата находятся в стоблце days_employed, это как раз и есть медианная замена. Данные все разные, это разные клиенты. Если я их удалю, то анализ будет искаженный. Поясни, пожалуйста.
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера_v2❌:</b> Не совсем - duplicated() работает так: он находит две (или более) дублирующихся строк, одной из них присваивает значение False (так как одну из дублирующихся записей нам надо оставить), остальным присваивает True - так как это повторения первой строки. В таблице выше ты видишь только те строки, которым он присвоил True, но у каждой из них есть двойник с значением False. Эти строки все же лучше удалить. 

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Николай, спасибо, я разобралась как работает метод duplicated()!
Убрала дубликаты

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v3✅:</b> Теперь согласен👍 </div>

### Шаг 2.5. Формирование дополнительных датафреймов словарей, декомпозиция исходного датафрейма.

Создадим два новых датафрейма.

Каждому уникальному значению из `family_status` соответствует уникальное значение `family_status_id`

In [32]:
df_family_status = df[['family_status', 'family_status_id']]
display(df_family_status.head(10)) # выведим данные для просмотра

family_status  family_status_id
0   женат / замужем                 0
1   женат / замужем                 0
2   женат / замужем                 0
3   женат / замужем                 0
4  гражданский брак                 1
5  гражданский брак                 1
6   женат / замужем                 0
7   женат / замужем                 0
8  гражданский брак                 1
9   женат / замужем                 0

Проверим словарь на наличие дубликатов

In [33]:
df_family_status.duplicated().sum()

21447

Выявили дубликаты. Методом `drop_duplicates()` удалим дубликаты с изменением индексов и удалением старых.

In [34]:
df_family_status = df_family_status.drop_duplicates().reset_index(drop=True)
display(df_family_status)

family_status  family_status_id
0        женат / замужем                 0
1       гражданский брак                 1
2         вдовец / вдова                 2
3              в разводе                 3
4  Не женат / не замужем                 4

Каждому уникальному значеию из `education` соответствует уникальное значение `education_id`

In [35]:
df_education = df[['education', 'education_id']]
display(df_education.head(10)) # выведим данные для просмотра

education  education_id
0    высшее             0
1   среднее             1
2   среднее             1
3   среднее             1
4   среднее             1
5    высшее             0
6    высшее             0
7   среднее             1
8    высшее             0
9   среднее             1

Проверим словарь на наличие дубликатов

In [36]:
df_education.duplicated().sum()

21447

Выявили дубликаты. Методом `drop_duplicates()` удалим дубликаты с изменением индексов и удалением старых.

In [37]:
df_education = df_education.drop_duplicates().reset_index(drop=True)
display(df_education)

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

Удалим из исходного датафрейма столбцы `education` и `family_status`, оставим только их идентификаторы: `education_id` и `family_status_id`.

In [38]:
df_total = df.drop(['education', 'family_status'], axis='columns') # применим метод drop()
display(df_total) # выведим данные для просмотра

children  days_employed  dob_years  education_id  family_status_id  \
0             1    8437.673028       42.0             0                 0   
1             1    4024.803754       36.0             1                 0   
2             0    5623.422610       33.0             1                 0   
3             3    4124.747207       32.0             1                 0   
4             0  340266.072047       53.0             1                 1   
...         ...            ...        ...           ...               ...   
21447         1    4529.316663       43.0             1                 1   
21448         0  343937.404131       67.0             1                 0   
21449         1    2113.346888       38.0             1                 1   
21450         3    3112.481705       38.0             1                 0   
21451         2    1984.507589       40.0             1                 0   

      gender income_type  debt  total_income                       purpose  
0          F   сотрудник     0        253875                 покупка жилья  
1          F   сотрудник     0        112080       приобретение автомобиля  
2          M   сотрудник     0        145885                 покупка жилья  
3          M   сотрудник     0        267628    дополнительное образование  
4          F   пенсионер     0        158616               сыграть свадьбу  
...      ...         ...   ...           ...                           ...  
21447      F   компаньон     0        224791             операции с жильем  
21448      F   пенсионер     0        155999          сделка с автомобилем  
21449      M   сотрудник     1         89672                  недвижимость  
21450      M   сотрудник     1        244093  на покупку своего автомобиля  
21451      F   сотрудник     0         82047         на покупку автомобиля  

[21452 rows x 10 columns]

Проверим таблицу на наличие дубликатов.

In [39]:
df_total.duplicated().sum()

0

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Цель этого пункта - сформировать таблицы, в которых будут лежать уникальные связки id с значениями типа образования и семейного статуса.

Чтобы сделать значения в твоих таблицах-словарях уникальными нужно удалить из них дубликаты</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Здесь речь идет о дубликах выше? Если нет, то прошу поясни, а то я не понимаю, что нужно поменять.
</div>

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера_v2❌:</b> Речь вот о чем - нашей задачей было сформировать таблицы - словари. В них должны лежать уникальные связки id и значения. Сейчас строки в словарях дублируюся - это видно из следующей ячейки; Дубликаты из словарей надо удалить</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Дубликаты в словарях удалила
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v3✅:</b> Отлично👍

In [40]:
df_education

education  education_id
0               высшее             0
1              среднее             1
2  неоконченное высшее             2
3            начальное             3
4       ученая степень             4

### Шаг 2.6. Построение сводной таблицы по количеству детей.

Создадим сводную таблицу, где методом `pivot_table()` сгруппируем данные по столбцу `children`, возьмем значения из столбца `debt`. Где в столбце `children` количество детей от 0 до 5, в столбце `0` количество клиентов вернувших кредит в срок, в столбце `1` количество клиентов не вернувших кредит в срок.

In [41]:
df_pivot_children = df_total.pivot_table(index=['children'], columns='debt', values='education_id', aggfunc='count')
display(df_pivot_children) # выведим данные для просмотра

debt            0       1
children                 
0         13026.0  1063.0
1          4410.0   445.0
2          1926.0   202.0
3           303.0    27.0
4            37.0     4.0
5             9.0     NaN

Методом `reset_index()` убирем мультииндекс.

In [42]:
df_pivot_children_reset_index = df_pivot_children.reset_index()
display(df_pivot_children_reset_index)

debt  children        0       1
0            0  13026.0  1063.0
1            1   4410.0   445.0
2            2   1926.0   202.0
3            3    303.0    27.0
4            4     37.0     4.0
5            5      9.0     NaN

Обнаружили, что в стоблце `1` (количество клиентов не вернувших кредит в срок) появилось значение `NaN`. Заменим `NaN` на 0 методом `fillna()`

In [43]:
df_pivot_children_reset_index[1] = df_pivot_children_reset_index[1].fillna(0)

Посчитаем процент клиентов, не выплатиших кредит в срок. Для этого создадим столбец `default rate`, где разделим не выплативших клиентов на общее количество клиентов.

In [44]:
df_pivot_children_reset_index['default rate'] = (df_pivot_children_reset_index[1]/(df_pivot_children_reset_index[0]+df_pivot_children_reset_index[1]))*100
display(df_pivot_children_reset_index) # выведим данные для просмотра

debt  children        0       1  default rate
0            0  13026.0  1063.0      7.544893
1            1   4410.0   445.0      9.165808
2            2   1926.0   202.0      9.492481
3            3    303.0    27.0      8.181818
4            4     37.0     4.0      9.756098
5            5      9.0     0.0      0.000000

<div class="alert alert-block alert-warning">
<b>Комментарий ревьюера⚠️:</b> Тут скорее не категоризация, а построение сводной таблицы

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Поправила название
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> 👍 </div>

### Шаг 2.6. Категоризация семейного положения.

Создадим сводную таблицу, где методом `pivot_table()` сгруппируем данные по столбцу `family_status_id`, возьмем значения из столбца `debt`. Где в столбце `family_status_id` семейное положение клиентов, в столбце `0` количество клиентов вернувших кредит в срок, в столбце `1` количество клиентов не вернувших кредит в срок.

In [45]:
df_pivot_family = df_total.pivot_table(index=['family_status_id'], columns='debt', values='education_id', aggfunc='count')
display(df_pivot_family) # выведим данные для просмотра

debt                  0    1
family_status_id            
0                 11408  931
1                  3761  388
2                   896   63
3                  1110   85
4                  2536  274

Методом `reset_index()` убирем мультииндекс.

In [46]:
df_pivot_family_reset_index = df_pivot_family.reset_index()
display(df_pivot_family_reset_index) # выведим данные для просмотра

debt  family_status_id      0    1
0                    0  11408  931
1                    1   3761  388
2                    2    896   63
3                    3   1110   85
4                    4   2536  274

Посчитаем процент клиентов, не выплатиших кредит в срок. Для этого создадим столбец `default rate`, где разделим не выплативших клиентов на общее количество клиентов.

In [47]:
df_pivot_family_reset_index['default rate'] = (df_pivot_family_reset_index[1]/(df_pivot_family_reset_index[0]+df_pivot_family_reset_index[1]))*100
display(df_pivot_family_reset_index) # выведим данные для просмотра

debt  family_status_id      0    1  default rate
0                    0  11408  931      7.545182
1                    1   3761  388      9.351651
2                    2    896   63      6.569343
3                    3   1110   85      7.112971
4                    4   2536  274      9.750890

Для наглядности добавим столбец с названиями семейного статуса. Для этого возьмем созданный датафрейм `df_family_status` и методом `drop_duplicates()` отфильтруем статусы.

In [48]:
df_family_status_drop = df_family_status.drop_duplicates()

Добавим полученную таблицу `df_family_status_drop` к таблице `df_pivot_family_reset_index` методом `merge()`.

In [49]:
df_pivot_family_reset_index_merge = df_pivot_family_reset_index.merge(df_family_status_drop, on='family_status_id', how = 'left')

In [50]:
display(df_pivot_family_reset_index_merge) # выведим данные для просмотра

family_status_id      0    1  default rate          family_status
0                 0  11408  931      7.545182        женат / замужем
1                 1   3761  388      9.351651       гражданский брак
2                 2    896   63      6.569343         вдовец / вдова
3                 3   1110   85      7.112971              в разводе
4                 4   2536  274      9.750890  Не женат / не замужем

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> 👍

### Шаг 2.6. Категоризация дохода.

Для удобства анализа создадим категории доходов:
* 0 – 30 000 — 'E';
* 30 001 – 50 000 — 'D';
* 50 001 – 200 000 — 'C';
* 200 001 – 1 000 000 — 'B';
* 1 000 001 и выше — 'A'.

Создадим функцию `income_group` на основании данных диапазонов.

In [51]:
def income_group(total_income): 
    if total_income  <= 30000:
        return 'E'
    if total_income >=30001 and total_income <=50000:
        return 'D'
    if total_income >= 50001 and total_income <=200000:
        return 'C'
    if total_income >= 200001 and total_income <=1000000:
        return 'B'
    return 'A'

Создадим столбец `total_income_category`, где методом `apply()` добавим категории функции `income_group` к таблице `df_total`.

In [52]:
df_total['total_income_category'] = df_total['total_income'].apply(income_group)
display(df_total.head(10)) # выведим данные для просмотра

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028       42.0             0                 0      F   
1         1    4024.803754       36.0             1                 0      F   
2         0    5623.422610       33.0             1                 0      M   
3         3    4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   
5         0     926.185831       27.0             0                 1      M   
6         0    2879.202052       43.0             0                 0      F   
7         0     152.779569       50.0             1                 0      M   
8         2    6929.865299       35.0             0                 1      F   
9         0    2188.756445       41.0             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category  
0                     B  
1                     C  
2                     C  
3                     B  
4                     C  
5                     B  
6                     B  
7                     C  
8                     C  
9                     C

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Функция написана верно, круто что результат ее выполнения проверен👍

Создадим сводную таблицу, где методом `pivot_table()` сгруппируем данные по столбцу `total_income_category`, возьмем значения из столбца `debt`. Где в столбце `total_income_category` категории доходов клиентов, в столбце `0` количество клиентов вернувших кредит в срок, в столбце `1` количество клиентов не вернувших кредит в срок.

In [53]:
df_pivot_total_income = df_total.pivot_table(index=['total_income_category'], columns='debt', values='education_id', aggfunc='count')
display(df_pivot_total_income) # выведим данные для просмотра

debt                       0     1
total_income_category             
A                         23     2
B                       4685   356
C                      14654  1360
D                        329    21
E                         20     2

Методом `reset_index()` убирем мультииндекс.

In [54]:
df_pivot_total_income_reset_index = df_pivot_total_income.reset_index()
display(df_pivot_total_income_reset_index) # выведим данные для просмотра

debt total_income_category      0     1
0                        A     23     2
1                        B   4685   356
2                        C  14654  1360
3                        D    329    21
4                        E     20     2

Посчитаем процент клиентов, не выплатиших кредит в срок. Для этого создадим столбец `default rate`, где разделим не выплативших клиентов на общее количество клиентов.

In [55]:
df_pivot_total_income_reset_index['default rate'] = (df_pivot_total_income_reset_index[1]/(df_pivot_total_income_reset_index[0]+df_pivot_total_income_reset_index[1]))*100
display(df_pivot_total_income_reset_index) # выведим данные для просмотра

debt total_income_category      0     1  default rate
0                        A     23     2      8.000000
1                        B   4685   356      7.062091
2                        C  14654  1360      8.492569
3                        D    329    21      6.000000
4                        E     20     2      9.090909

### Шаг 2.7. Категоризация целей кредита.

Для удобства анализа создадим следующие категории:
* операции с автомобилем,
* операции с недвижимостью,
* проведение свадьбы,
* получение образования.

Создадим функцию `purpose_group`, которая сгруппирует по категориям цели кредита.

In [56]:
def purpose_group(purpose): 
    if 'авт' in purpose:
        return 'операции с автомобилем'
    if 'жиль' in purpose:
        return 'операции с недвижимостью'
    if 'недвиж' in purpose:
        return 'операции с недвижимостью'
    if 'свад'in purpose:
        return 'проведение свадьбы'
    if 'образов' in purpose:
        return 'получение образования'

Создадим столбец `purpose_category`, где методом `apply()` добавим категории функции `purpose_group` к таблице `df_total`.

In [57]:
df_total['purpose_category'] = df_total['purpose'].apply(purpose_group)
display(df_total.head(10)) # выведим данные для просмотра

children  days_employed  dob_years  education_id  family_status_id gender  \
0         1    8437.673028       42.0             0                 0      F   
1         1    4024.803754       36.0             1                 0      F   
2         0    5623.422610       33.0             1                 0      M   
3         3    4124.747207       32.0             1                 0      M   
4         0  340266.072047       53.0             1                 1      F   
5         0     926.185831       27.0             0                 1      M   
6         0    2879.202052       43.0             0                 0      F   
7         0     152.779569       50.0             1                 0      M   
8         2    6929.865299       35.0             0                 1      F   
9         0    2188.756445       41.0             1                 0      M   

  income_type  debt  total_income                     purpose  \
0   сотрудник     0        253875               покупка жилья   
1   сотрудник     0        112080     приобретение автомобиля   
2   сотрудник     0        145885               покупка жилья   
3   сотрудник     0        267628  дополнительное образование   
4   пенсионер     0        158616             сыграть свадьбу   
5   компаньон     0        255763               покупка жилья   
6   компаньон     0        240525           операции с жильем   
7   сотрудник     0        135823                 образование   
8   сотрудник     0         95856       на проведение свадьбы   
9   сотрудник     0        144425     покупка жилья для семьи   

  total_income_category          purpose_category  
0                     B  операции с недвижимостью  
1                     C    операции с автомобилем  
2                     C  операции с недвижимостью  
3                     B     получение образования  
4                     C        проведение свадьбы  
5                     B  операции с недвижимостью  
6                     B  операции с недвижимостью  
7                     C     получение образования  
8                     C        проведение свадьбы  
9                     C  операции с недвижимостью

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> Отличная функция, результат ее выполнения проверен, молодец👍

Создадим сводную таблицу, где методом `pivot_table()` сгруппируем данные по столбцу `purpose_category`, возьмем значения из столбца `debt`. Где в столбце `purpose_category` категории целей кредита, в столбце `0` количество клиентов вернувших кредит в срок, в столбце `1` количество клиентов не вернувших кредит в срок.

In [58]:
df_pivot_purpose = df_total.pivot_table(index=['purpose_category'], columns='debt', values='education_id', aggfunc='count')
display(df_pivot_purpose) # выведим данные для просмотра

debt                          0    1
purpose_category                    
операции с автомобилем     3903  403
операции с недвижимостью  10028  782
получение образования      3643  370
проведение свадьбы         2137  186

Методом `reset_index()` убирем мультииндекс.

In [59]:
df_pivot_purpose_reset_index = df_pivot_purpose.reset_index()
display(df_pivot_purpose_reset_index) # выведим данные для просмотра

debt          purpose_category      0    1
0       операции с автомобилем   3903  403
1     операции с недвижимостью  10028  782
2        получение образования   3643  370
3           проведение свадьбы   2137  186

Посчитаем процент клиентов, не выплатиших кредит в срок. Для этого создадим столбец `default rate`, где разделим не выплативших клиентов на общее количество клиентов.

In [60]:
df_pivot_purpose_reset_index['default rate'] = (df_pivot_purpose_reset_index[1]/(df_pivot_purpose_reset_index[0]+df_pivot_purpose_reset_index[1]))*100
display(df_pivot_purpose_reset_index) # выведим данные для просмотра

debt          purpose_category      0    1  default rate
0       операции с автомобилем   3903  403      9.359034
1     операции с недвижимостью  10028  782      7.234043
2        получение образования   3643  370      9.220035
3           проведение свадьбы   2137  186      8.006888

### Ответы на вопросы.

##### Вопрос 1:  Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [61]:
display(df_pivot_children_reset_index) # выведим данные для просмотра

debt  children        0       1  default rate
0            0  13026.0  1063.0      7.544893
1            1   4410.0   445.0      9.165808
2            2   1926.0   202.0      9.492481
3            3    303.0    27.0      8.181818
4            4     37.0     4.0      9.756098
5            5      9.0     0.0      0.000000

Клиенты, у которых 5 детей, возвращают кредит в срок. Но так как таких клиентов всего 9 человек из 21524 (что составляет 0,04% и является не существенным для анализа), то эти данные не будем учитывать для итогового вывода. 

Полученные данные:
* Процент клиентов должников, у которых нет детей составляет 7,51%;
* Процент клиентов должников, у которых 1 ребенок составляет 9,15%;
* Процент клиентов должников, у которых 2 ребенка составляет 9,48%;
* Процент клиентов должников, у которых 3 детей составляет 8,18%;
* Процент клиентов должников, у которых 4 детей составляет 9,76%.

##### Вывод 1:

Клиенты у которых нет детей возвращают кредит в срок чаще, чем клиенты у которых есть дети. 

##### Вопрос 2:  Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [62]:
display(df_pivot_family_reset_index_merge) # выведим данные для просмотра

family_status_id      0    1  default rate          family_status
0                 0  11408  931      7.545182        женат / замужем
1                 1   3761  388      9.351651       гражданский брак
2                 2    896   63      6.569343         вдовец / вдова
3                 3   1110   85      7.112971              в разводе
4                 4   2536  274      9.750890  Не женат / не замужем

Полученные данные:
* Процент клиентов должников, состоящих в браке составляет 7,52%;
* Процент клиентов должников, состоящих в разводе составляет 7,11%;
* Процент клиентов должников, состоящих в гражданском браке составляет 9,29%;
* Процент клиентов должников, не состоящих в браке составляет 9,74%;
* Процент клиентов должников, имеющие семейный статус вдовец/вдова составляет 6,56%.

##### Вывод 2:

Чаще всего возвращают кредиты в срок клиенты, имеющие статус вдовец/вдова, а также клиенты состоящие в браке и разводе. А клиенты не состоящие в браке или состоящие в гражданском браке реже возвращают в кредит в срок. 

##### Вопрос 3:  Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [63]:
display(df_pivot_total_income_reset_index) # выведим данные для просмотра

debt total_income_category      0     1  default rate
0                        A     23     2      8.000000
1                        B   4685   356      7.062091
2                        C  14654  1360      8.492569
3                        D    329    21      6.000000
4                        E     20     2      9.090909

Полученные данные:
* Процент клиентов должников, имеющий доход от 30 000 до 50 000 составляет 6%;
* Процент клиентов должников, имеющий доход от 200 000 до 1 000 000 составляет 7,06%;
* Процент клиентов должников, имеющий доход от 1 000 000 и выше составляет 8%;
* Процент клиентов должников, имеющий доход от 50 000 до 200 000 составляет 8,45%;
* Процент клиентов должников, имеющий доход до 30 000 составляет 9,09%.

##### Вывод 3:

Чаще всего возвращают кредиты в срок клиенты, имеющие доход от 30 000 до 50 000, реже всего клиенты, имеющие доход до 30 000.

##### Вопрос 4: Как разные цели кредита влияют на его возврат в срок?

In [64]:
display(df_pivot_purpose_reset_index) # выведим данные для просмотра

debt          purpose_category      0    1  default rate
0       операции с автомобилем   3903  403      9.359034
1     операции с недвижимостью  10028  782      7.234043
2        получение образования   3643  370      9.220035
3           проведение свадьбы   2137  186      8.006888

Полученные данные:
* Процент клиентов должников, взявшие кредит на операции с недвижимостью составляет 7,21%;
* Процент клиентов должников, взявшие кредит на проведение свадьбы составляет 7,92%;
* Процент клиентов должников, взявшие кредит на операции с автомобилями составляет 9,34%;
* Процент клиентов должников, взявшие кредит на получение образования составляет 9,20%.

##### Вывод 4:

Чаще всего возвращают кредиты в срок клиенты, взявшие кредиты на операции с недвижимостью и на проведение свадьбы, реже всего клиенты, взявшие кредит на операции с автомобилями и получение образования.

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> Прошу поправить оформление на этом шаге. Перед ответом на каждый вопрос пожалуйста добавь вывод сводной таблицы, которые мы строили ранее

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b> С ответами согласен

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Добавила таблицы к каждому выводу
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> 👍 </div>

## Общий вывод:

Мы искали закономерности и установили:

1.  Есть ли зависимость между количеством детей и возвратом кредита в срок?

Клиенты у которых нет детей возвращают кредит в срок чаще, чем клиенты у которых есть дети.

2.  Есть ли зависимость между семейным положением и возвратом кредита в срок?

Чаще всего возвращают кредиты в срок клиенты, имеющие статус вдовец/вдова, а также клиенты состоящие в браке и разводе. А клиенты не состоящие в браке или состоящие в гражданском браке реже возвращают в кредит в срок. 

3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Чаще всего возвращают кредиты в срок клиенты, имеющие доход от 30 000 до 50 000, реже всего клиенты, имеющие доход до 30 000.

4. Как разные цели кредита влияют на его возврат в срок?

Чаще всего возвращают кредиты в срок клиенты, взявшие кредиты на операции с недвижимостью и на проведение свадьбы, реже всего клиенты, взявшие кредит на операции с автомобилями и получение образования.

В результате исследования были обнаружены закономерности, которые можно учесть при построении модели кредитного скоринга, которая оценит способность потенциального заёмщика вернуть кредит банку.

<div class="alert alert-block alert-danger">
<b>Комментарий ревьюера❌:</b> В конце проекта принято кратко описывать все проделанные шаги и полученные результаты. Зачем это нужно - когда проект захочет посмотреть будущий работодатель, у него может не быть времени на подробный разбор кода. Вероятнее всего он бегло просмотрит код, но захочет изучить результат, который будет в общем выводе. Поэтому все же советую дополнить общий вывод)
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b> Отличный вывод, теперь работа логчески завершена и мы можем понять суть исследования🔥 </div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера✅:</b>Мне очень понравилась твоя работа. Все рассуждения верные, очень логичные и осмысленные. Ничего лишнего, всё по полкам. Ты подробно разбираешься в деталях, для будущего аналитика это очень важная черта😊

Проект еще нужно немного доработать, а именно:
    
- исправить блок удаления дубликатов
    
- исправить блок формирования таблиц-словарей
    
- исправить блок ответов на вопросы
    
- исправить общий вывод
    
Остальные замечания оставляю на твое усмотрение😉

Жду твою работу на повторное ревью.
    
[Тут есть полезные приемы Markdown](https://paulradzkov.com/2014/markdown_cheatsheet/)

</div>

<div class="alert alert-block alert-info">
<b>Комментарий студента:</b> 
Николай, спасибо, что помогаешь делать проект лучше!
Исправила общий вывод.
Помоги, пожалуйста, разобраться в дубликатах и я доделаю проект.    
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v2✅:</b>
Ты отлично справилась с доработками, осталось поправить дубликаты и таблицы словари)
    
С остальным все хорошо🔥
    
Жду твою работу на повторное ревью.
</div>

<div class="alert alert-block alert-success">
<b>Комментарий ревьюера_v3✅:</b> Отличная работа, проект принимаю. Могу с уверенностью сказать - ты хорошо владеешь пройденным материалом. Мне понравилась твоя реакция на обратную связь, комментарии даже к желтым замечаниям. Желаю удачи в дальнейшей учебе🔥 </div>